Exercises/Dynamic Memory/Exercise5/Exercise5-9.cpp

```diff
@@ -0,0 +1,197 @@
+* The program will halt execution right before the memory is freed. Use the memory viewer to see
+* the memory location of your student object. Once you have that view, press F10 to execute the
+* line where you free the memory. You�ll see that the memory where your object lived has
+* changed to dd. This is a debug compiler code that means it�s uninitialized heap memory, so you
+* can be certain the delete worked properly.
+* Now, run your program again. This time, go to the memory location of the name of the student.
+* Once you�ve found it, press F10 to delete the object. You�ll see that the memory has not been
+* freed.
+* Print to the console why deleting the student object does not delete the name�s char array.
+* 
+* 
+* 
+* Exercise 8 - Destructors and Dynamic Allocation 2
+* Keeping in mind that deleting an object calls its destructor right before freeing the memory, and
+* using the knowledge you have about destructors, update the program to fix this memory leak.
+* You may not delete the memory allocated for the name of a student from outside the student
+* object.
+* 
+* 
+* 
+* Exercise 9 - Destructors and Dynamic Allocation 3
+* Great, we�ve fixed the memory leak! But memory leaks are sneaky, and there�s another
+* scenario we�ve failed to take into consideration. We created a setName method where we�re
+* allocating memory for the student�s name, that�s good and all, but what happens if you call that
+* method twice to update the name a second time. The original memory is not being released in
+* that case.
+* How would you fix this?
+* At this point, it�s obvious that keeping track of all memory allocations in a program can be a
+* hassle. In the future, we�ll discuss techniques that will make life a lot easier for us.
+*/
+
+
+
+#include <iostream>
+#include <string>
+
+class FStudent
+{
+
+private:
+	char* Name;
+	int Age;
+
+public:
+	
+	//--------------------------------------------------------------
+	FStudent()
+	{
+		Name = nullptr;
+		Age = 0;
+	}
+
+	//--------------------------------------------------------------
+	FStudent(int InAge)
+	{
+		Name = nullptr;
+		Age = InAge;
+	}
+
+	//--------------------------------------------------------------
+	~FStudent()
+	{
+		delete Name;
+	}
+
+	//--------------------------------------------------------------
+	void SetName(const char* NewName)
+	{
+		if (Name)
+		{
+			delete Name;
+			Name = nullptr;
+		}
+
+		int NameLength = 0;
+		for (int i = 0; NewName[i] != '\0'; i++)
+		{
+			NameLength++;
+		}
+
+		Name = new char[NameLength+1];
+		memcpy_s(Name, sizeof(char) * (NameLength+1), NewName, sizeof(char) * (NameLength+1));
+	}
+
+	//--------------------------------------------------------------
+	void SetAge(const int NewAge)
+	{
+		Age = NewAge;
+	}
+
+	//--------------------------------------------------------------
+	void Print()
+	{
+		std::cout << "Name: " << Name << std::endl;
+		std::cout << "Age: " << Age << std::endl;
+	}
+};
+
+
+
+int main()
```

Comment:
This method is quite big, you could refactor it by extracting methods :)

https://refactoring.guru/smells/long-method